sdk/communication/azure-communication-chat/azure/communication/chat/_shared/user_credential.py

```diff
@@ -3,84 +3,139 @@
+        if not isinstance(token, six.string_types):
+            raise TypeError("Token must be a string.")
+        self._token = create_access_token(token)
+        self._token_refresher = kwargs.pop('token_refresher', None)
+        self._proactive_refresh = kwargs.pop('proactive_refresh', False)
+        if(self._proactive_refresh and self._token_refresher is None):
+            raise ValueError("When 'proactive_refresh' is True, 'token_refresher' must not be None.")
+        self._timer = None
         self._lock = Condition(Lock())
         self._some_thread_refreshing = False
+        self._is_closed = Event()
 
     def get_token(self, *scopes, **kwargs):  # pylint: disable=unused-argument
         # type (*str, **Any) -> AccessToken
         """The value of the configured token.
         :rtype: ~azure.core.credentials.AccessToken
         """
 
-        if not self._token_refresher or not self._token_expiring():
+        if not self._token_refresher or not self._is_token_expiring_soon(self._token):
             return self._token
 
-        should_this_thread_refresh = False
+        self._update_token_and_reschedule()
+        return self._token
 
+    def _update_token_and_reschedule(self):
+        should_this_thread_refresh = False
         with self._lock:
-            while self._token_expiring():
+            while self._is_token_expiring_soon(self._token):
                 if self._some_thread_refreshing:
-                    if self._is_currenttoken_valid():
+                    if self._is_token_valid(self._token):
                         return self._token
-
-                    self._wait_till_inprogress_thread_finish_refreshing()
+                    self._wait_till_lock_owner_finishes_refreshing()
                 else:
                     should_this_thread_refresh = True
                     self._some_thread_refreshing = True
                     break
 
         if should_this_thread_refresh:
             try:
-                newtoken = self._token_refresher()  # pylint:disable=not-callable
-
+                new_token = self._token_refresher()
+                if not self._is_token_valid(new_token):
+                    raise ValueError(
+                        "The token returned from the token_refresher is expired.")
                 with self._lock:
-                    self._token = newtoken
+                    self._token = new_token
                     self._some_thread_refreshing = False
                     self._lock.notify_all()
             except:
                 with self._lock:
                     self._some_thread_refreshing = False
                     self._lock.notify_all()
-
                 raise
+        if self._proactive_refresh:
+            self._schedule_refresh()
         return self._token
 
-    def _wait_till_inprogress_thread_finish_refreshing(self):
+    def _schedule_refresh(self):
+        if self._is_closed.is_set():
+            return
```

Comment:
This check was added for pypy issue when the canceled thread tries to reschedule another timer, and for that reason, I think we do not have to throw an exception

Reply:
Yeah I think that makes sense.
I'm just wondering if someone calls `credential.close()` then attempts to reuse the credential object without a context manager - this would silently ignore the proactive refresh.
But maybe there's a better place to check for that.

Reply:
Yeah, I understood your point, I think the good place to check is the get_token method, what do you think?

Reply:
That could work - though it would probably need to be within the `if self._proactive_refresh` clause.